core/client/hammer/hammer.go

```diff
@@ -76,125 +97,199 @@ func New(ctx context.Context, dial DialFunc, callOptions CallOptions,
 		timeout:     timeout,
 		ktCli:       ktCli,
 		appID:       fmt.Sprintf("hammer_%v", time.Now().Format("2006-01-02/15:04:05")),
-		config:      config,
+		domain:      domain,
 
 		signers:        []*tink.KeysetHandle{keyset},
 		authorizedKeys: authorizedKeys.Keyset(),
 	}, nil
 }
 
-// Run runs operationCount with up to maxWorkers over ramp duration.
-func (h *Hammer) Run(ctx context.Context, operationCount, maxWorkers int, ramp time.Duration) {
-	jobs := gen(ctx, operationCount)
-	outputs := h.launchWorkers(ctx, jobs, maxWorkers, ramp)
-	h.printStats(1*time.Second, outputs)
-}
+// Run runs a total of operationCount operations across numWorkers.
+// The number of workers should roughly be (goal QPS / timeout seconds)
+func (h *Hammer) Run(ctx context.Context, numWorkers int, c Config) error {
+	workers, err := h.newWorkers(numWorkers)
+	if err != nil {
+		return err
+	}
+
+	/*
+		bars := make([]*cpb.ProgressBar, 0, len(workers))
+		for _, w := range workers {
+			bars = append(bars, w.bar)
+		}
+		pool, err := cpb.StartPool(bars...)
+		if err != nil {
+			return err
+		}
+	*/
 
-// launchWorkers gradually adds workers to the worker pool over ramp up to workers.
-func (h *Hammer) launchWorkers(ctx context.Context, jobs <-chan int, workers int, ramp time.Duration) <-chan error {
-	outputs := make(chan error)
+	if ok := c.TestTypes["batch"]; ok {
+		// Batch Write users
+		log.Printf("Batch Write")
+		args := genArgs(ctx, c.BatchWriteQPS, c.BatchWriteSize, c.BatchWriteCount, c.Duration)
+		handlers := make([]ReqHandler, 0, len(workers))
+		for i := range workers {
+			handlers = append(handlers, workers[i].writeOp)
+		}
+		log.Printf("workers: %v", len(handlers))
+		executeRequests(ctx, args, handlers)
+		fmt.Print("\n")
```

Comment:
What's the purpose of this new line? Could it be put together with "Batch Write" above? Note that the dots printed by `executeRequests` are likely to appear after this `Print`.

Reply:
The purpose of the new line is to separate the dots between the request types.

`executeRequests` blocks until the workers have finished. 
Are you sure that it will be an issue?

Reply:
Oh, it blocks indeed, sorry.

Hm, so if I understand correctly, the hammer will do multiple types of requests, but not simultaneously: first all batch writes, then all single writes, then all reads. Right? In this case I don't see why we can't have a slightly simpler hammer that accepts only one operation type (instead of a comma-separated list), and just run this binary multiple times. (there are couple more advantages of that, see my previous review messages)